Add new fields to status/dom_sensor/pm tables in STATE_DB for CMIS/C-CMIS

longhuan-cisco opened the pull request:
Add new fields to status/dom_sensor/pm tables in STATE_DB for CMIS/C-CMIS

#### Description
1. Add all the fields that are supported by c_cmis [get_transceiver_bulk_status](https://github.com/sonic-net/sonic-platform-common/blob/b26aa5ef79fca95064d0baf3b1beed8d56d212f8/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py#L302) API but not currently present to dom_sensor table.
2. Add new fields (provided by [cmis.py:get_transceiver_status() API](https://github.com/sonic-net/sonic-platform-common/blob/86bab38c723eb4ebbfa16feed66344d1b3ffd46e/sonic_platform_base/sonic_xcvr/api/public/cmis.py#L1548) and [c_cmis.py:get_transceiver_status() API](https://github.com/sonic-net/sonic-platform-common/blob/86bab38c723eb4ebbfa16feed66344d1b3ffd46e/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py#L565) ) to status table(TRANSCEIVER_STATUS). 
following SW fields(non-eeprom-fetched) are already present in today's status table:
```
status                       = 1*255VCHAR                       ; code of the module status (plug in, plug out)
error                        = 1*255VCHAR                       ; module error (N/A or a string consisting of error descriptions joined by "|", like "error1 | error2" )
```
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;following HW fields (eeprom-fetched) are added per [this HLD](https://github.com/sonic-net/SONiC/blob/a5675ea78a7cb6d4d1dfdc780fbf2183d7a41a89/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#214-transceiver-status-table) (this is just snippet, refer to HLD for complete list):
```
module_state                 = 1*255VCHAR                       ; current module state (ModuleLowPwr, ModulePwrUp, ModuleReady, ModulePwrDn, Fault)
module_fault_cause           = 1*255VCHAR                       ; reason of entering the module fault state
datapath_firmware_fault      = BOOLEAN                          ; datapath (DSP) firmware fault
module_firmware_fault        = BOOLEAN                          ; module firmware fault
module_state_changed         = BOOLEAN                          ; module state changed
datapath_hostlane1           = 1*255VCHAR                       ; data path state indicator on host lane 1
......
rxsigpowerhighwarning_flag   = BOOLEAN                          ; rxsigpower high warning flag
rxsigpowerlowwarning_flag    = BOOLEAN                          ; rxsigpower low warning flag
```
3. Add pm related fields to a newly created table TRANSCEIVER_PM in STATE_DB. (DB schema is defined in this [HLD](https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#215-transceiver-pm-table)
4. Add corresponding DB update/delete handling along the xcvrd flow.
5. Add/modify test cases on test_xcvrd.py to cover newly added code.

Here's the entire DB schema of [status tbl](https://github.com/sonic-net/SONiC/blob/a5675ea78a7cb6d4d1dfdc780fbf2183d7a41a89/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#214-transceiver-status-table) and [dom_sensor tbl](https://github.com/sonic-net/SONiC/blob/a5675ea78a7cb6d4d1dfdc780fbf2183d7a41a89/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#212-transceiver-dom-sensor-table), according to the new HLD from PR https://github.com/sonic-net/SONiC/pull/1076


#### Motivation and Context
<!--
     Why is this change required? What problem does it solve?
     If this pull request closes/resolves an open Issue, make sure you
     include the text "fixes #xxxx", "closes #xxxx" or "resolves #xxxx" here
-->

#### How Has This Been Tested?
<!--
     Please describe in detail how you tested your changes.
     Include details of your testing environment, and the tests you ran to
     see how your change affects other areas of the code, etc.
-->
Verified on CMIS/C-CMIS optics, interfaces come up fine, and new fields show up in DB

UT output for dom_sensor table:
[dom_sensor_tbl_ut_output.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9805375/dom_sensor_tbl_ut_output.txt)

UT output for status table:
[status_tbl_ut_output.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9805349/status_tbl_ut_output.txt)

UT output for status table in xcvrd restart case:
[xcvrd_restart_test.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9892550/xcvrd_restart_test.txt)

UT output for pm table:
[pm_table_output.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9957822/pm_table_output.txt)

UT output for pm table in xcvrd restart case:
[xcvrd_restart_test_pm_tbl.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9957826/xcvrd_restart_test_pm_tbl.txt)

UT result for test_xcvrd.py:
[ut_w_coverage.txt](https://github.com/sonic-net/sonic-platform-daemons/files/9957863/ut_w_coverage.txt)


#### Additional Information (Optional)

> **Note**
> Please don't merge this PR until the dependency PR https://github.com/sonic-net/sonic-platform-common/pull/315 gets merged.

longhuan-cisco commented:
@prgeor please review

mihirpat1 reviewed (commented):
Can you please also test restarting xcvrd to ensure the DB gets deleted and updated accordingly

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -856,19 +840,65 @@ def waiting_time_compensation_with_sleep(time_start, time_to_wait):
+                return SFP_EEPROM_NOT_READY
+
+        except NotImplementedError:
+            helper_logger.log_error("This functionality is currently not implemented for this platform")
```
Added unique log.

longhuan-cisco commented:
> Can you please also test restarting xcvrd to ensure that the table gets deleted and updated accordingly

Yea Attached restart test log in description.

longhuan-cisco commented:
@prgeor @mihirpat1 I took care of the existing comments, can you please review?

longhuan-cisco commented:
Added pm table related change. 
Changed this PR's title to track changes for all three tables: status/dom_sensor/pm. (Since they all touch the similar flow in xcvrd)

prgeor commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -507,60 +538,47 @@ def post_port_dom_info_to_db(logical_port_name, port_mapping, table, stop_event=
+                pm_info_dict = _wrapper_get_transceiver_pm(physical_port)
+                if pm_info_cache is not None:
+                    # If cache is enabled, put dom information to cache
+                    pm_info_cache[physical_port] = pm_info_dict
```
where are we going to use this pm_info_cache[physica_port] ? I dont see it being used

longhuan-cisco replied:
pm_info_cache is passed to post_port_pm_info_to_db() from task_worker() of ```class DomInfoUpdateTask``` (in task_worker, we already have similar cache logic for dom_info dom_th_info in today's code)
pm_info_cache is used here:
```  
        if pm_info_cache is not None and physical_port in pm_info_cache:
            # If cache is enabled and pm info is in cache, just read from cache, no need read from EEPROM
            pm_info_dict = pm_info_cache[physical_port]

``` 
One possible situation I can think of: task_worker() later walks through a different logical port which belongs to a previously visited physical port, then cache is hit.

keboliu commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -856,19 +867,65 @@ def waiting_time_compensation_with_sleep(time_start, time_to_wait):
+
+def delete_port_from_status_table_hw(logical_port_name, port_mapping, status_tbl):
+    for physical_port_name in get_physical_port_name_dict(logical_port_name, port_mapping).values():
+        found, fvs = status_tbl.get(physical_port_name)
```
why use **physical_port_name** as the index and not **logical_port_name**?

longhuan-cisco replied:
Because this function is supposed to only delete the HW fields of status tbl, which are fetched from eeprom (more precisely VDM) and per physical port.

keboliu replied:
I believe most of the current xcvrd relevant fileds in the state DB are fetching from the hw(EEPROM), but all of them are using  'logical_port_name' as the index, there will be duplicated information in the DB(multiple logical_port can be created from the same physical port), but the user can easily query them from the CLI/SNMP with the logical port name. I suggest to follow the same convention here.  @prgeor please also share your view here.

longhuan-cisco replied:
AFAIK, SW part (fields 'status' and 'error') of status table are not from EEPROM, and all the other fields on status table (introduced by this PR) and other tables are from EEPROM. 
And currently, only SW part of status table is using logical_port_name as key, all the other existing tables (dom_sensor tbl/info tbl, those are eeprom-feteched) are using physical_port_name as index. I'm just making the HW part of status table aligned with other existing tables.

e.g. in today’s post_port_sfp_info_to_db(), post_port_dom_threshold_info_to_db() and post_port_dom_info_to_db(), they all have the following code to use physical port to set DB:
    for physical_port in physical_port_list:
        …..
        port_name = get_physical_port_name(logical_port_name, ganged_member_num, ganged_port)
        …..
        table.set(port_name, fvs)

We discussed with @prgeor about this before, and we agreed to better use logical_port_name as key, and https://github.com/sonic-net/sonic-buildimage/issues/10144 can be used to track this pending work item. But the plan was to keep this item as lower priority, since it may only cause issue on DPB case for now. And also the history/necessarity of ganged port needs to figured out first (ganged port can have 1:n mapping from logical port to physical port), so that we can decide wether the ganged port related logic can be cleaned up around this area.

keboliu replied:
> AFAIK, SW part (fields 'status' and 'error') of status table are not from EEPROM, and all the other fields on status table (introduced by this PR) and other tables are from EEPROM. And currently, only SW part of status table is using logical_port_name as key, all the other existing tables (dom_sensor tbl/info tbl, those are eeprom-feteched) are using physical_port_name as index. I'm just making the HW part of status table aligned with other existing tables.

> 
> e.g. in today’s post_port_sfp_info_to_db(), post_port_dom_threshold_info_to_db() and post_port_dom_info_to_db(), they all have the following code to use physical port to set DB:
> 
> ```
>     for physical_port in physical_port_list:
>         …..
>         port_name = get_physical_port_name(logical_port_name, ganged_member_num, ganged_port)
>         …..
>         table.set(port_name, fvs)
> ```
>

This is not true, check below code you will see that for all none-ganged ports it will always use logical port name including table DOM and Transceiver info table. By the way, I haven't seen any vendor claiming that they have ganged port on any of their platforms.

```
def get_physical_port_name(logical_port, physical_port, ganged):
    if ganged:
        return logical_port + ":{} (ganged)".format(physical_port)
    else:
        return logical_port
```
 
> We discussed with @prgeor about this before, and we agreed to better use logical_port_name as key, and [sonic-net/sonic-buildimage#10144](https://github.com/sonic-net/sonic-buildimage/issues/10144) can be used to track this pending work item to clean up all the existing code to use logical_port_name as key for DB tables. But the plan was to keep this item as lower priority, since it may only cause issue on DPB case for now. And also the history/necessarity of ganged port needs to be figured out first (ganged port can have 1:n mapping from logical port to physical port), so that we can decide wether the ganged port related logic can be cleaned up around this area. Also, to clean this area up, more consideration/discussion would be needed later on whether we need to divide the corresponding eeprom-fetched fields belonging to the same physical port into multiple groups based on logical ports. e.g. for the following fields in TRANSCEIVER_INFO table, if we are in breakout mode, each logical port will only have part of the 8 lanes, we might need to divide them based on logical port.
>

If you change the index of the tables the CLIs (e.g 'show interface transceiver eeprom') will be broken, and also SNMP and other telemetry applications assuming the index shall be logical port index will also be broken, please double confirm.
 
> ```
>         active_apsel_hostlane1       = INTEGER                  ; active application selected code assigned to host lane 1
>         active_apsel_hostlane2       = INTEGER                  ; active application selected code assigned to host lane 2
>         active_apsel_hostlane3       = INTEGER                  ; active application selected code assigned to host lane 3
>         active_apsel_hostlane4       = INTEGER                  ; active application selected code assigned to host lane 4
>         active_apsel_hostlane5       = INTEGER                  ; active application selected code assigned to host lane 5
>         active_apsel_hostlane6       = INTEGER                  ; active application selected code assigned to host lane 6
>         active_apsel_hostlane7       = INTEGER                  ; active application selected code assigned to host lane 7
>         active_apsel_hostlane8       = INTEGER                  ; active application selected code assigned to host lane 8
> ```

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -2124,6 +2212,8 @@ def retry_eeprom_reading(self):
                 post_port_dom_info_to_db(logical_port, self.port_mapping, self.xcvr_table_helper.get_dom_tbl(asic_index))
                 post_port_dom_threshold_info_to_db(logical_port, self.port_mapping, self.xcvr_table_helper.get_dom_tbl(asic_index))
+                update_port_transceiver_status_table_hw(logical_port, self.port_mapping, self.xcvr_table_helper.get_status_tbl(asic_index))
+                post_port_pm_info_to_db(logical_port, self.port_mapping, self.xcvr_table_helper.get_pm_tbl(asic_index))
```
Inside post_port_pm_info_to_db(), the following statement will make sure to skip for other non-ZR cases (where platform API get_transceiver_pm() throws NotImplementedError and then _wrapper_get_transceiver_pm() returns empty dict):
```
            if not pm_info_dict:
                continue
```

longhuan-cisco commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -507,60 +538,47 @@ def post_port_dom_info_to_db(logical_port_name, port_mapping, table, stop_event=
 
         except NotImplementedError:
-            helper_logger.log_error("This functionality is currently not implemented for this platform")
+            helper_logger.log_error("get_transceiver_pm is currently not implemented for this platform")
```
Yes, you are right. Removed it.

prgeor commented on sonic-xcvrd/xcvrd/xcvrd.py:
```diff
@@ -507,62 +538,44 @@ def post_port_dom_info_to_db(logical_port_name, port_mapping, table, stop_event=
+                continue
+            beautify_pm_info_dict(pm_info_dict, physical_port)
+            fvs = swsscommon.FieldValuePairs([(k, v) for k, v in pm_info_dict.items()])
+            table.set(physical_port_name, fvs)
```
can we keep the key as logical port just like done for DOM and other table?

prgeor replied:
This is because the CLI and streaming telemetry will need logical port as key in future when we do breakout of physical port to many logical ports.

prgeor replied:
@keboliu if you check the usage of port_name in previous change:-

table.set(port_name, fvs) 

port_name is nothing but physical port name at line 496 in previous code :- 

port_name = get_physical_port_name(logical_port_name, ganged_member_num, ganged_port)

so, i don't see this PR causing any issue. We need to revisit once we support breakout in Xcvrd.

prgeor commented:
@StormLiangMS please help cherry-pick to 202211

StormLiangMS commented:
@prgeor I suppose you want to have this one in 202211 other than 202111
